blant.c

```diff
@@ -1,1976 +1,1977 @@
-		nextChar++;
-		for(orbit=0; orbit < _numOrbits; orbit++)
-		{
-		    assert(isdigit(*nextChar));
-		    numRead = sscanf(nextChar, "%lu", &count);
-		    assert(numRead == 1);
-		    ODV(lineNum,orbit) += count;
-		    while(isdigit(*nextChar)) nextChar++; // read past current integer
-		    assert(*nextChar == ' ' || (orbit == _numOrbits-1 && *nextChar == '\n'));
-		    nextChar++;
-		}
-		assert(*nextChar == '\0');
-		break;
-	    case indexGraphlets: case indexOrbits:
-		fputs(line, stdout);
-		if(_windowRep) 
-			while(fgets(line, sizeof(line), fpThreads[thread])) 
-				fputs(line, stdout);
-		break;
-	    default:
-		Abort("oops... unknown _outputMode in RunBlantInThreads while reading child process");
-		break;
-	    }
-	}
-	lineNum++;
-    } while(!done);
-    for(i=0; i<_THREADS; i++)
-	fclose(fpThreads[i]);  // close all the pipes
-
-    // if numSamples is not a multiple of _THREADS, finish the leftover samples
-    int leftovers = numSamples % _THREADS;
-    return RunBlantFromGraph(_k, leftovers, G);
-}
-
-void BlantAddEdge(int v1, int v2)
-{
-    if(!_pairs) _pairs = Malloc(2*_maxEdges*sizeof(_pairs[0]));
-    assert(_numEdges <= _maxEdges);
-    if(_numEdges >= _maxEdges)
-    {
-	_maxEdges *=2;
-	_pairs = Realloc(_pairs, 2*_maxEdges*sizeof(int));
-    }
-    _numNodes = MAX(_numNodes, v1+1); // add one since, for example, if we see a node numbered 100, numNodes is 101.
-    _numNodes = MAX(_numNodes, v2+1);
-    _pairs[2*_numEdges] = v1;
-    _pairs[2*_numEdges+1] = v2;
-    if(_pairs[2*_numEdges] == _pairs[2*_numEdges+1])
-	Fatal("BlantAddEdge: edge %d (%d,%d) has equal nodes; cannot have self-loops\n", _numEdges, v1, v2);
-    if(_pairs[2*_numEdges] > _pairs[2*_numEdges+1])
-    {
-	int tmp = _pairs[2*_numEdges];
-	_pairs[2*_numEdges] = _pairs[2*_numEdges+1];
-	_pairs[2*_numEdges+1] = tmp;
-    }
-    assert(_pairs[2*_numEdges] < _pairs[2*_numEdges+1]);
-    _numEdges++;
-}
-
-int RunBlantEdgesFinished(int k, int numSamples, int numNodes, char **nodeNames)
-{
-    GRAPH *G = GraphFromEdgeList(_numNodes, _numEdges, _pairs, SPARSE);
-    Free(_pairs);
-    _nodeNames = nodeNames;
-    return RunBlantInThreads(k, numSamples, G);
-}
-
-// Initialize the graph G from an edgelist; the user must allocate the pairs array
-// to have 2*numEdges elements (all integers), and each entry must be between 0 and
-// numNodes-1. The pairs array MUST be allocated using malloc or calloc, because
-// we are going to free it right after creating G (ie., before returning to the caller.)
-int RunBlantFromEdgeList(int k, int numSamples, int numNodes, int numEdges, int *pairs)
-{
-    assert(numNodes >= k);
-    GRAPH *G = GraphFromEdgeList(numNodes, numEdges, pairs, SPARSE);
-    Free(pairs);
-    return RunBlantInThreads(k, numSamples, G);
-}
-
-const char const * const USAGE = \
-
-    "USAGE: blant [-r seed] [-t threads (default=1)] [-m{outputMode}] [-d{displayMode}] {-n nSamples | -c confidence -w width} {-k 
```

Comment:
Rather than declaring an int pointer and malloc'ing one lousy int, make piped a normal int, and then when you want to pass it by "reference", by passing it's pointer; eg closeFile(fpGraph, &piped)